pennylane/plugins/default_qubit.py

```diff
@@ -213,7 +221,42 @@ def mat_vec_product(self, mat, vec, wires):
         perm = wires + unused_idxs
         inv_perm = np.argsort(perm)  # argsort gives inverse permutation
         state_multi_index = np.transpose(tdot, inv_perm)
-        return np.reshape(state_multi_index, 2 ** self.num_wires)
+
+        return state_multi_index
+
+    def mat_vec_product_einsum(self, mat, vec, wires):
+        r"""Apply multiplication of a matrix to subsystems of the quantum state.
+
+        This function uses einsum instead of tensordot. This approach is only
+        faster for single- and two-qubit gates.
+
+        Args:
+            mat (array): matrix to multiply
+            vec (array): state tensor to multiply
+            wires (Sequence[int]): target subsystems
+
+        Returns:
+            array: output vector after applying ``mat`` to input ``vec`` on specified subsystems
+        """
+        mat = np.reshape(mat, [2] * len(wires) * 2)
+
+        state_indices = ABC[: self.num_wires]
+        affected_indices = "".join(ABC_ARRAY[wires].tolist())
+        new_indices = ABC[self.num_wires : self.num_wires + len(wires)]
```

Comment:
(Minor suggestion) Perhaps a comment like this could be helpful here
```suggestion
        # Indices of mat which remain as new indices in the state after contraction
        new_indices = ABC[self.num_wires : self.num_wires + len(wires)]
```